statistics/handle/statscache.go

```diff
@@ -15,15 +15,271 @@ package handle
+	// since newRistrettoStatsCache controls the memory usage by itself, set the capacity of
+	// the underlying LRUCache to max to close its memory control
+	sc := &ristrettoStatsCache{
+		memCapacity: memoryLimit,
+		memTracker:  memory.NewTracker(memory.LabelForStatsCache, -1),
+	}
+	for i := range sc.tablesShards {
+		sc.tablesShards[i].data = make(map[int64]*statistics.Table)
+	}
+	sc.version = 0
+	cache, err := cache.NewCache(&cache.Config{
+		NumCounters: 1e7,         // number of keys to track frequency of (10M).
+		MaxCost:     memoryLimit, // maximum cost of cache (1GB).
+		BufferItems: 64,          // number of keys per Get buffer.
+		Cost: func(value interface{}) int64 {
+			return (value.(*statistics.Table).MemoryUsage())
+		},
+		OnEvict: func(key uint64, value interface{}) {
+			if t, ok := sc.Get(int64(key)); ok {
+				if value != nil && t.PhysicalID == int64(key) {
+					sc.memTracker.Consume(-t.MemUsage)
+					sc.Set(int64(key), t.CopyWithoutBucketsAndCMS())
+				}
+			}
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	sc.cache = cache
+	return sc, nil
+}
+
+// BytesConsumed returns the consumed memory usage value in bytes.
+// only used by test.
+func (sc *ristrettoStatsCache) BytesConsumed() int64 {
+	rndTbl := sc.GetAll()[0]
+	sc.Erase(rndTbl.PhysicalID)
+	// insert a table memory = 0 to make cache renew the memory
+	for i := 0; i < 10; i++ {
+		sc.Insert(rndTbl)
+		time.Sleep(10 * time.Millisecond)
+	}
+	return sc.memTracker.BytesConsumed()
+}
+
+// GetBytesLimit get the limits of memory.
+func (sc *ristrettoStatsCache) GetBytesLimit() int64 {
+	return sc.memTracker.GetBytesLimit()
+}
+
+// GetMutex return the Muetex
+func (sc *ristrettoStatsCache) GetMutex() *sync.Mutex {
+	return &sc.mu
+}
+
+// SetBytesLimit set new byteslimit
+func (sc *ristrettoStatsCache) SetBytesLimit(bytesLimit int64) {
+	sc.cache.SetNewMaxCost(bytesLimit)
+}
+
+// Close close the cache.
+func (sc *ristrettoStatsCache) Close() {
+	sc.cache.Close()
+}
+
+// Clear clears the cache data.
+func (sc *ristrettoStatsCache) Clear() {
+	sc.cache.Clear()
+	sc.mu.Lock()
+	sc.version = 0
+	sc.mu.Unlock()
+	for i := range sc.tablesShards {
+		sc.tablesShards[i].Lock()
+		sc.tablesShards[i].data = make(map[int64]*statistics.Table)
+		sc.tablesShards[i].Unlock()
+	}
+	sc.memTracker = memory.NewTracker(memory.LabelForStatsCache, -1)
+
+}
+
+// GetAll get all the tables.
+func (sc *ristrettoStatsCache) GetAll() []*statistics.Table {
+	tables := make([]*statistics.Table, 0)
+	for i := range sc.tablesShards {
+		shard := &sc.tablesShards[i]
+		for _, tbl := range shard.data {
+			ntbl, _ := sc.Lookup(tbl.PhysicalID)
+			tables = append(tables, ntbl)
+		}
+	}
+	return tables
+}
+
+// lookupUnsafe get table with id without Lock.
+func (sc *ristrettoStatsCache) lookupUnsafe(id int64) (*statistics.Table, bool) {
+	key := uint64(id)
+	value, hit := sc.cache.Get(key)
+	if !hit || value == nil || value.(*statistics.Table).PhysicalID != id {
```

Comment:
When the condition `value.(*statistics.Table).PhysicalID != id` will be true?

Reply:
i 'll remove this.